xdsl/transforms/experimental/function_constant_pinning.py

```diff
@@ -83,17 +83,19 @@ def match_and_rewrite(self, func_op: func.FuncOp, rewriter: PatternRewriter, /):
         # grab a reference to the next operation in the remainder.
         # this is because we will modify the op and therefore loose the "old" next op.
         next_op = function_remainder.next_op
+
         # unless we already hit the block terminator
-        if next_op is not None:
-            # while we haven't reached the return statement:
-            while function_remainder is not func_op.body.block.last_op:
-                # detatch the function
-                function_remainder.detach()
-                # re-insert it inside the else block of the if statement
-                rewriter.insert_op(function_remainder, InsertPoint.at_end(dest_block))
-                # go to next op
-                function_remainder = next_op
-                next_op = function_remainder.next_op
+        # while we haven't reached the return statement:
+        while (
+            next_op is not None and function_remainder is not func_op.body.block.last_op
+        ):
+            # detatch the function
+            function_remainder.detach()
+            # re-insert it inside the else block of the if statement
+            rewriter.insert_op(function_remainder, InsertPoint.at_end(dest_block))
+            # go to next op
+            function_remainder = next_op
+            next_op = function_remainder.next_op
```

Comment:
Why?!?!?!?!?!?

Reply:
this one actually makes sense to me, there's no way for the type system to know that just because `function_remainder` is not the last op that `next_op` will not be none, if `function_remainder.next_op` has type `Operation | None`